### DISCUSSION TITLE
Changing PWM pin on attiny404

### DISCUSSION THREAD

**Discussion by nhansendev** [SpenceKonde/megaTinyCore#1034] [Q&A] [answered]
I'm simply trying to generate a 1kHz PWM signal of variable duty cycle on pin PA4 of an attiny404.

After following the TCA [documentation](https://ww1.microchip.com/downloads/en/Appnotes/TB3217-Getting-Started-with-TCA-DS90003217.pdf) from Microchip I think I have everything set, but there's no output on PA4.

Is this right so far? I'm skeptical about my use of PORTMUX.CTRLC:
```
#define F_CPU 3660000UL
#include <avr/io.h>

#define PWM_FREQ 1024 // Hz

double PWM_const = F_CPU/(8*PWM_FREQ);

void initPWM(void){
  // Set up Timer/Counter Type A (TCA)

  // Set pin directions
  PORTA.DIR = 0b11010000;

  // Direct PWM output to pin PA4 (WO4)
  PORTMUX.CTRLC = PORTMUX_TCA04_bm;

  // Set to compare channel 0 and bottom Dual-Slope PWM mode
  TCA0.SINGLE.CTRLB = TCA_SINGLE_CMP0EN_bm | TCA_SINGLE_WGMODE_DSBOTTOM_gc;

  // Count on tick, not event (this is already the default)
  TCA0.SINGLE.EVCTRL &= !(TCA_SINGLE_CNTEI_bm);

  // https://ww1.microchip.com/downloads/en/Appnotes/TB3217-Getting-Started-with-TCA-DS90003217.pdf
  // Set the frequency of the PWM signal to 1024 Hz
  TCA0.SINGLE.PERBUF = uint16_t(PWM_const);

  // Set the duty cycle to 50%
  TCA0.SINGLE.CMP0BUF = uint16_t(0.5*PWM_const);

  // Set the prescaler and start the timer
  // Note: changing the prescaler changes the PWM frequency too
  TCA0.SINGLE.CTRLA = TCA_SINGLE_CLKSEL_DIV4_gc | TCA_SINGLE_ENABLE_bm;

}

int main(void)
{
    initPWM();
    
    while (1) {     
    }        
}

```

**Comment by hmeijdam** [comment]
When I look in the datasheet I see no WO4 associated with Control C 
WO0 to WO3 are available.
![image](https://github.com/SpenceKonde/megaTinyCore/assets/9595465/4734fcd8-87fc-486e-a48b-097dadaaaaaa)

But keep in mind you cannot select any pin for WO0 (which is what you are working with). 
According the I/O multiplexing table for WO0 you can choose from PB0 (default) and PB3(alternate).
![image](https://github.com/SpenceKonde/megaTinyCore/assets/9595465/d5f5943c-6f9b-42bf-bb2c-6370e7262c87)

Below sketch will get you 1024 Hz  on the alternate pin PB3

I have no 3.6 MHz crystal, so have changed the F_CPU logic to work with the standard Megatinycore clocks.

```
// #define F_CPU 3660000UL
#include <avr/io.h>
#define PWM_FREQ 1024 // Hz

double PWM_const = F_CPU/(8*PWM_FREQ);

void initPWM(void){
  // Set up Timer/Counter Type A (TCA)

  // Set pin directions
  // PORTA.DIR = 0b11010000;

  VPORTB.DIR |= PIN3_bm;// set PB3 as output

  // Direct PWM Wo0 output to it's alternate pin (PB3)
  PORTMUX.CTRLC = PORTMUX_TCA00_bm;

  // Set to compare channel 0 and bottom Dual-Slope PWM mode
  TCA0.SINGLE.CTRLB = TCA_SINGLE_CMP0EN_bm | TCA_SINGLE_WGMODE_DSBOTTOM_gc;

  // Count on tick, not event (this is already the default)
  TCA0.SINGLE.EVCTRL &= !(TCA_SINGLE_CNTEI_bm);

  // https://ww1.microchip.com/downloads/en/Appnotes/TB3217-Getting-Started-with-TCA-DS90003217.pdf
  // Set the frequency of the PWM signal to 1024 Hz
  TCA0.SINGLE.PERBUF = uint16_t(PWM_const);

  // Set the duty cycle to 50%
  TCA0.SINGLE.CMP0BUF = uint16_t(0.5*PWM_const);

  // Set the prescaler and start the timer
  // Note: changing the prescaler changes the PWM frequency too
  TCA0.SINGLE.CTRLA = TCA_SINGLE_CLKSEL_DIV4_gc | TCA_SINGLE_ENABLE_bm;
}

int main(void)
{  F_CPU_init ();   // reconfigure CPU clock prescaler
    initPWM();
    
    while (1) {     
    }        
}

void F_CPU_init () {
  // reconfigure CPU clock prescaler
#if (F_CPU == 20000000)
  /* No division on clock */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
#elif (F_CPU == 16000000)
  /* No division on clock */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, 0x00);
#elif (F_CPU == 10000000) // 20MHz prescaled by 2
  /* Clock DIV2 */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
#elif (F_CPU == 8000000) // 16MHz prescaled by 2
  /* Clock DIV2 */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_2X_gc));
#elif (F_CPU == 5000000) // 20MHz prescaled by 4
  /* Clock DIV4 */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_4X_gc));
#elif (F_CPU == 4000000) // 16MHz prescaled by 4
  /* Clock DIV4 */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_4X_gc));
#elif (F_CPU == 1000000) // 16MHz prescaled by 16
  /* Clock DIV16 */
  _PROTECTED_WRITE(CLKCTRL_MCLKCTRLB, (CLKCTRL_PEN_bm | CLKCTRL_PDIV_16X_gc));
#else
#ifndef F_CPU
#error "F_CPU not defined"
#else
#error "F_CPU defined as an unsupported value for untuned internal oscillator"
#endif
#endif
}
```

**Reply by nhansendev** [reply]
Thanks for the detailed response.

To clarify then, what can WO4 be used for? I must be misinterpreting the multiplexing table since it shows WO4 in the TCA0 column of PA4 (from the [attiny404 datasheet](https://ww1.microchip.com/downloads/en/devicedoc/50002687a.pdf)):
![image](https://github.com/SpenceKonde/megaTinyCore/assets/9289200/982a0132-6bf5-4b79-b32c-8d0ad07f18f4)

Edit: I see what you mean now. Looking at the CTRLC options only allows WO0 to switch between the two predefined pins.
![image](https://github.com/SpenceKonde/megaTinyCore/assets/9289200/688b4ded-736b-4bc5-ae71-01f0696c8349)

Then what is the purpose of WO4?

**Reply by grandaspanna** [reply, marked as answer]
You need to be in "split mode" to use WO3-WO5, which the core does by default. Limits resolution to 8 bits, which may or may not impact your needs.

The guide you linked does have examples on split mode.

I think if you use the PWM in the core, it supports output on PA4 by default. There are some other threads here on how to tweak frequency/duty cycle in the core.

**Reply by nhansendev** [reply]
This was the solution, thank you. In the future I'll just use the single mode with the understanding of the pin limitations, but for now this is what I changed the PWM setup to:

```
void initPWM(void){
  // Set up Timer/Counter Type A (TCA)

  // Enable split mode (needed for using PA4 as output)
  TCA0.SPLIT.CTRLD = TCA_SPLIT_SPLITM_bm;

  // Set compare channel 0 for high 1, which is the address for WO4 (PA4)
  TCA0.SPLIT.CTRLB = TCA_SPLIT_HCMP1EN_bm;

  // Set frequency
  TCA0.SPLIT.HPER = uint16_t(PWM_const);

  // Set the duty cycle (placeholder)
  TCA0.SPLIT.HCMP1 = uint16_t(0.5*PWM_const);

  // Set the prescaler and enable
  TCA0.SPLIT.CTRLA = TCA_SPLIT_CLKSEL_DIV4_gc | TCA_SPLIT_ENABLE_bm;  
}
```

**Reply by SpenceKonde** [reply]
Easier and cleaner: 

takeOverTCA0();
//TCA0 is now disabledm in single mode, and in the power on reset configuration, and you may configure it freely and the core will not consider it a pwm pin for analogWrite/digitalWrite

**Reply by nhansendev** [reply]
> takeOverTCA0();

Is there a link to documentation, or an example that you could share, please?
Evidently it's not as simple as calling the function, but I'm not sure what needs importing beyond `avr/io.h`

**Reply by SpenceKonde** [reply]
like always on Arduino, in the .ino file, and any library that includes Arduino.h (which it absolutely should) will get io.h automatically. 

And uh. It really is as simple as calling that function if you want full control over a TCA. Note that it does NOT alter portmux, so thjat must still be set right.

That will disable timer, and issue the hard reset to reset the TCA into the power on state (the default register values shown in in the datasheet, andanalogWrite won't try to turn on PWM (likely incorrectly) and digitalWrite wont try to turn it off (likely on the wrong pin. See the examples for the megaTinyCore "Library"

**Reply by nhansendev** [reply]
Ah, I wasn't clear before. I'm using PlatformIO instead of Arduino.
When I try it with just `io.h` imported it complains about `takeOverTCAO()` being undefined.
E.g.:
```
#include <avr/io.h>
takeOverTCAO(); // undefined
```

Edit: I see from some searching that PlatformIO is not directly supported, so I'll look at switching to Arduino
Thank you
